### PR TITLE
fix(pub): Support deserializing hosted deps without version constraint

### LINKED ISSUE #9310: Pub cannot parse Yaml file
A customer from us started to experience issues with Pub since some time:
```
IncorrectTypeException at dev_dependencies.lint on line 18, column 8: Expected element to be YamlMap but is YamlNull
	at com.charleskorn.kaml.YamlNodeKt.error(YamlNode.kt:278)
	at com.charleskorn.kaml.YamlNodeKt.getYamlMap(YamlNode.kt:272)
	at org.ossreviewtoolkit.plugins.packagemanagers.pub.DependencyMapSerializer.decodeDependency(Pubspec.kt:128)
```

```
08:53:13.076 [DefaultDispatcher-worker-1] ERROR org.ossreviewtoolkit.plugins.packagemanagers.pub.Pub - Failed to parse pubspec.yaml for package functional_listener:2.3.1: IncorrectTypeException: Expected element to be YamlMap but is YamlNull
```
The pub file `/home/nino/.pub-cache/hosted/pub.dev/functional_listener-2.3.1/pubspec.yaml`:
```
dev_dependencies:
  flutter_test:
    sdk: flutter
  lint:
```
Maybe in `YamlNode.decodeDependency()` should check if this is a map ?

Bug happening with ORT 34.0.0.

### PR DESCRIPTION
**Pull request by fviernau**
Fixes #9310.

### CODE REVIEW

**Review by nnobelis — APPROVED** [review]

File: plugins/package-managers/pub/src/main/kotlin/model/Pubspec.kt [hunk 1]
```diff
@@ -71,7 +72,7 @@ internal data class Pubspec(
     /** See https://dart.dev/tools/pub/dependencies#hosted-packages. */
     @Serializable
     data class HostedDependency(
-        val version: String,
+        val version: String?,
```
Comment (sschuberth) [inline plugins/package-managers/pub/src/main/kotlin/model/Pubspec.kt:75]: For consistency with `url`, shouldn't this also have a default value of `null`?

Note that if this class would be used with regular deserialization, and not with a custom serializer only, an explicit default value of `null` would even be required, as KxS in contrast to Jackson does not assign `null` to nullable types that are missing from the input by default.

File: plugins/package-managers/pub/src/main/kotlin/model/Pubspec.kt [hunk 2]
```diff
@@ -113,6 +114,9 @@ private object DependencyMapSerializer : KSerializer<Map<String, Dependency>> by
 
     private fun YamlNode.decodeDependency(): Dependency {
         if (this is YamlScalar) return HostedDependency(yamlScalar.content)
+        // https://dart.dev/tools/pub/dependencies says that "The version constraint is optional but recommended. If no
+        // version constraint is given, any is assumed.".
```
Comment (sschuberth) [inline plugins/package-managers/pub/src/main/kotlin/model/Pubspec.kt:118]: Shouldn't this better go to documentation for the property in the class?

Comment (fviernau) [reply plugins/package-managers/pub/src/main/kotlin/model/Pubspec.kt:118]: I've made another PR: https://github.com/oss-review-toolkit/ort/pull/9313

File: plugins/package-managers/pub/src/main/kotlin/model/Pubspec.kt (hunk 1 above)
Comment (fviernau) [reply plugins/package-managers/pub/src/main/kotlin/model/Pubspec.kt:75]: I've made another PR: https://github.com/oss-review-toolkit/ort/pull/9313

**Pull request merged**